packages/gatsby/src/utils/worker/render-html.js

```diff
@@ -7,7 +7,8 @@ const generatePathToOutput = outputPath => {
   let outputFileName = outputPath.replace(/^(\/|\\)/, ``) // Remove leading slashes for webpack-dev-server
 
   if (!/\.(html?)$/i.test(outputFileName)) {
-    outputFileName = path.join(outputFileName, `index.html`)
+    outputFileName =
+      outputFileName + `${outputFileName.endsWith(`/`) ? `index` : ``}.html`
```

Comment:
I think this is dangerous - it relies on hosting matching `/path/something` to `/path/something.html` which I'm not sure how common is (probably less common than matching index.html). This change could potentially break reactjs.org (which uses .html paths) without people knowing

Reply:
Do you have any ideas on how we could test this? Are you just thinking this is a difference in OS's potentially?

Reply:
Ah, I see your point. Do you have any other idea to avoid the 301 redirect? For example, loading this page triggers it: https://reactjs.org/languages

Reply:
> Do you have any ideas on how we could test this?

Other than trying it out on top X of hosting services/software (or reading docs for them), I see no other way :(

> Ah, I see your point. Do you have any other idea to avoid the 301 redirect?

Ideally we would only change `gatsby serve` here and not the output that Gatsby produces right now (that change have breaking potential). Ideally we could configure `express.static` to not do redirects, but last time I checked this wasn't possible. We might need to:
 1) make feature request to [`serve-static`](https://expressjs.com/en/resources/middleware/serve-static.html) to make 301 behaviour configurable (so it use regular 200 response instead of 301 for those paths)
 2) vendor `serve-static` and apply patch we need ourselves?
 3) or look for something that will not do 301 instead of `serve-static`?

(`serve-static` is package used by `express.static`)

Reply:
also, fact that we need to change `express.static` config to support new format is what I worry about with this change

Reply:
I understand the implications... And I see why my changes are too dangerous.

What about... looping through the public folder instead of using express.static()? Pseudo-code:
```
const serve = (dir = './public') => {
  const folders = listFolders(dir);
  const files = listFiles(dir);

  for each folder in folders {
    serve(dir + folder);
  }

  [custom logic to serve HTML files without trailing slash]
}
```

Reply:
Few additional alternatives:

1. We can add "200 redirect" ourselves in express middleware before we register `express.static`:

```js
  router.use((req, res, next) => {
    if (!/.+\..+$/g.test(req.url) && req.url.slice(-1) !== '/') {
      req.url = `${req.url}/`;
    }
    next();
  })
```
this would "200 redirect non-slash non-extension requests"

Problem is that those might be actual files :/

2. We can use `redirect: false` (and maybe `fallthrough: true`) in option for `express.static` and add new route handler after `express.static` to try to send file on our own?

```js
  router.use((req, res, next) => {
    if (!/.+\..+$/g.test(req.url) && req.url.slice(-1) !== '/') {
      // let's try adding trailing `/index.html`, see if file exist and send it if it does
    } else {
      next();
    }
  })
```

Reply:
This is a very good idea. I just updated the PR, and also updated the test repository, it works well! Just re-install node_modules and run patch-package to test it.